### PR TITLE
Multithreading in perl

### PR DESCRIPTION
**Pull request by henningonsbring**
I have done a template that can be used as a starting point when multithreading in perl. The template is a script that can run as it is so you will see that it works before you modify the code to do what you want.

How to view changes made to manual?

```
git clone -b <branch name> https://github.com/Clinical-Genomics/development.git
cd development
conda create -n D_mkdocs
source activate D_mkdocs
pip install -r requirements.txt
mkdocs serve
```

go to your browser: http://127.0.0.1:8000

Publishing the changes when you are done:
```
mkdocs gh-deploy
```

### CODE REVIEW

File: multithreading-in-perl/README.md [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+
+    #!/usr/bin/perl
+    # Template for multithreading in perl
+
+    use strict;
+    use warnings;
+    use threads;
+
+    my $n_threads = 4; # How many processes that will be done in parallel
+    my @thr = ();
```
Comment (henrikstranneheim) [inline multithreading-in-perl/README.md:13]:
```suggestion
    my @threads;
```

File: multithreading-in-perl/README.md [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+
+    # Loop that starts multiple tasks in parallel on separate threads
+    for (my $i = 0; $i < $n_threads; $i++) {
+        $thr[$i] = threads->create({'context' => 'list'}, \&generic_sub, $i);
+    }
```
Comment (henrikstranneheim) [inline multithreading-in-perl/README.md:18]:
```suggestion
    for my $thread_index (0 .. $MAX_THREADS) {
        $threads[$thread_index] = threads->create({'context' => 'list'}, \&generic_sub, $thread_index);
    }
```

File: multithreading-in-perl/README.md [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+The threads module in perl can be used to run multiple processes in parallel. This is an example of a script that use multithreading:
+
+    #!/usr/bin/perl
+    # Template for multithreading in perl
+
+    use strict;
+    use warnings;
+    use threads;
+
+    my $n_threads = 4; # How many processes that will be done in parallel
```
Comment (henrikstranneheim) [inline multithreading-in-perl/README.md:12]:
```suggestion
    # How many processes that will be done in parallel
    my $MAX_THREADS = 4;
```

File: multithreading-in-perl/README.md [hunk 4]
```diff
@@ -0,0 +1,42 @@
...
+    my @thr = ();
+
+    # Loop that starts multiple tasks in parallel on separate threads
+    for (my $i = 0; $i < $n_threads; $i++) {
+        $thr[$i] = threads->create({'context' => 'list'}, \&generic_sub, $i);
+    }
+
+    # To return calculations from threads
+    for (my $i = 0; $i < $n_threads; $i++) {
+        my ($returned_tid, $returned_sum) = $thr[$i]->join();
```
Comment (henrikstranneheim) [inline multithreading-in-perl/README.md:22]:
```suggestion
    for my $thread_index (0 .. $MAX_THREADS) {
        my ($returned_tid, $returned_sum) = $threads[thread_index]->join();
```

File: multithreading-in-perl/README.md [hunk 5]
```diff
@@ -0,0 +1,42 @@
...
+        print "The thread with thread ID " . $returned_tid . " got the sum " . $returned_sum . "\n";
```
Comment (henrikstranneheim) [inline multithreading-in-perl/README.md:23]:
```suggestion
        say {STDOUT} qq{The thread with thread ID $returned_tid  got the sum  $returned_sum};
```

File: multithreading-in-perl/README.md [hunk 6]
```diff
@@ -0,0 +1,42 @@
...
+    }
+
+    # Subroutine that is done multiple times in parallel
+    sub generic_sub {
+        my ($thread_id) = @_;
+        print "Thread " . $thread_id . " started" . "\n";
```
Comment (henrikstranneheim) [inline multithreading-in-perl/README.md:29]:
```suggestion
        say {STDOUT} qq{Thread $thread_id started};
```

File: multithreading-in-perl/README.md [hunk 7]
```diff
@@ -0,0 +1,42 @@
...
+        my $sum = 0;
+        for (my $i = 0; $i < 10000000; $i++) {
```
Comment (henrikstranneheim) [inline multithreading-in-perl/README.md:31]: Test to redo the loop according to my suggestions above

File: multithreading-in-perl/README.md [hunk 8]
```diff
@@ -0,0 +1,42 @@
...
+            $sum = $sum + rand();
+        }
+        print "Thread " . $thread_id . " done" . "\n";
```
Comment (henrikstranneheim) [inline multithreading-in-perl/README.md:34]: And modify the print to a say

**Review by henrikstranneheim — COMMENTED** [review]
Nice!

File: multithreading-in-perl/README.md (hunk 5 above)
Comment (henningonsbring) [reply multithreading-in-perl/README.md:24]: will really {STDOUT} work? I do {\*STDOUT} instead

File: multithreading-in-perl/README.md (hunk 6 above)
Comment (henningonsbring) [reply multithreading-in-perl/README.md:30]: I use {\\*STDOUT} here as well

Comment (henrikstranneheim) [reply multithreading-in-perl/README.md:30]: No, you are right I forgot the "*" It should be {*STDOUT}

File: multithreading-in-perl/README.md (hunk 8 above)
Comment (henningonsbring) [reply multithreading-in-perl/README.md:35]: After testing the script with `for f in {1..20}; do perl multi_thread.pl; done` I notice that there is an error that occur when say is used: "String found where operator expected". This error does not occur when print is used. This error does not occur every time the script is executed, so I guess there has to be some unlucky timing and a lot of processes try to do something with STDOUT at the same time?

File: multithreading-in-perl/README.md (hunk 6 above)
Comment (henningonsbring) [reply multithreading-in-perl/README.md:30]: {\*STDOUT} or {\\*STDOUT}?

**Review by henrikstranneheim — APPROVED** [review]
👍

**Comment by henrikstranneheim** [comment]
@henningonsbring Nice work! Shall I merge?

**Comment by henningonsbring** [comment]
Yes, let's merge this!

**Pull request merged**